lib/3scale_toolbox/remote_cache.rb

```diff
@@ -0,0 +1,153 @@
+module ThreeScaleToolbox
+  class RemoteCache < BasicObject
+
+    attr_reader :metrics_cache, :methods_cache, :backend_metrics_cache, :backend_methods_cache, :subject
+
+    def initialize(subject)
+      @subject = subject
+      # Metrics and methods cache data
+      @metrics_cache = {}
+      # methods cache data
+      @methods_cache = {}
+      # Backend Metrics and methods cache data
+      @backend_metrics_cache = {}
+      # Backend methods cache data
+      @backend_methods_cache = {}
+    end
+
+    def list_metrics(service_id)
+      return metrics_cache[service_id] if metrics_cache.has_key? service_id
+
+      subject.list_metrics(service_id).tap do |metrics|
+        metrics_cache[service_id] = metrics unless metrics.respond_to?(:has_key?) && !metrics['errors'].nil?
+      end
+    end
+
+    def list_methods(service_id, metric_id)
+      key = method_cache_key(service_id, metric_id)
+      return methods_cache[key] if methods_cache.has_key? key
+
+      subject.list_methods(service_id, metric_id).tap do |methods|
+        methods_cache[key] = methods unless methods.respond_to?(:has_key?) && !methods['errors'].nil?
+      end
+    end
+
+    def create_metric(service_id, attributes)
+      subject.create_metric(service_id, attributes).tap do |_|
+        metrics_cache.delete(service_id)
```

Comment:
When creating a metric, why is the metric being deleted from the cache and not overwritten it?
What happens with the possible error result on `create_metric`?

Reply:
When the existing cache is being invalidated because a metric is deleted or updated, the naive implementation is to delete the existing cache. Smarter implementation is postponed a good nice to have. But better in another PR.

Implemented error handling on create and update methods

Reply:
I understand about invalidating the cache when the metric is being deleted or updated and it makes total sense, but this specific case is about metric creation :thinking:

Reply:
The same for metric update, you can update your cached data with the updated metric. Can  be easily done. But not sure we want to enter into  this issue. Not for this PR. IF we implement caching update, the cache will never be invalidated for the entire command lifetime. I am not sure we should implement.

Reply:
:+1: